icu4c/source/test/cintltst/ulocbuildertst.c

```diff
@@ -0,0 +1,1743 @@
+        "0",  // Too short
+        "a",  // Too short
+        "123456789",  // Too long
+        "abcdefghi",  // Too long
+        "ab-123456789",  // Second subtag too long
+        "abcdefghi-12",  // First subtag too long
+        "a-ab-87654321",  // Third subtag too long
+        "87654321-a-0-3",  // First subtag too long
+    };
+    const char * aToZ = "abcdefghijklmnopqrstuvwxyz";
+    const int32_t aToZLen = static_cast<int32_t>(uprv_strlen(aToZ));
+    int32_t i = 0;
+    for (const char* ill : illFormed) {
+        char ch = aToZ[i];
+        i = (i + 1) % aToZLen;
+        UErrorCode status = U_ZERO_ERROR;
+        LocaleBuilder bld;
+        bld.setExtension(ch, ill);
+        Locale loc = bld.build(status);
+        if (status != U_ILLEGAL_ARGUMENT_ERROR) {
+            errln("setExtension('%c', \"%s\") should fail but has no Error\n",
+                  ch, ill);
+        }
+    }
+}
+
+static void TestSetLocale() {
+    LocaleBuilder bld1, bld2;
+    UErrorCode status = U_ZERO_ERROR;
+    Locale l1 = bld1.setLanguage("en")
+        .setScript("Latn")
+        .setRegion("MX")
+        .setVariant("3456-abcde")
+        .addUnicodeLocaleAttribute("456")
+        .addUnicodeLocaleAttribute("123")
+        .setUnicodeLocaleKeyword("nu", "thai")
+        .setUnicodeLocaleKeyword("co", "stroke")
+        .setUnicodeLocaleKeyword("ca", "chinese")
+        .build(status);
+    if (U_FAILURE(status) || l1.isBogus()) {
+        errln("build got Error: %s\n", u_errorName(status));
+    }
+    status = U_ZERO_ERROR;
+    Locale l2 = bld1.setLocale(l1).build(status);
+    if (U_FAILURE(status) || l2.isBogus()) {
+        errln("build got Error: %s\n", u_errorName(status));
+    }
+
+    if (l1 != l2) {
+        errln("Two locales should be the same, but one is '%s' and the other is '%s'",
+              l1.getName(), l2.getName());
+    }
+}
+
+static void TestPosixCases() {
+    UErrorCode status = U_ZERO_ERROR;
+    Locale l1 = Locale::forLanguageTag("en-US-u-va-posix", status);
+    if (U_FAILURE(status) || l1.isBogus()) {
+        errln("build got Error: %s\n", u_errorName(status));
+    }
+    LocaleBuilder bld;
+    bld.setLanguage("en")
+        .setRegion("MX")
+        .setScript("Arab")
+        .setUnicodeLocaleKeyword("nu", "Thai")
+        .setExtension('x', "1");
+    // All of above should be cleared by the setLocale call.
+    Locale l2 = bld.setLocale(l1).build(status);
+    if (U_FAILURE(status) || l2.isBogus()) {
+        errln("build got Error: %s\n", u_errorName(status));
+    }
+    if (l1 != l2) {
+        errln("The result locale should be the set as the setLocale %s but got %s\n",
+              l1.toLanguageTag<std::string>(status).c_str(),
+              l2.toLanguageTag<std::string>(status).c_str());
+    }
+    Locale posix("en-US-POSIX");
+    if (posix != l2) {
+        errln("The result locale should be the set as %s but got %s\n",
+              posix.getName(), l2.getName());
+    }
+}
+*/
```

Comment:
Why commented out?

Reply:
I am still working on them since the API is not yet approved.